crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -204,4 +257,42 @@ impl Context {
             _ => unimplemented!("The cast operation is only valid for integers."),
         }
     }
+
+    /// Returns the `AcirVar` that was previously stored at the given address.
+    fn convert_ssa_load(&mut self, address: &ValueId) -> AcirVar {
+        let (array_id, index) =
+            self.ssa_value_to_array_address.get(address).expect("ICE: Load from undeclared array");
+        self.acir_context.array_load(*array_id, *index).expect("invalid array load")
+    }
+
+    /// Returns true if the value has been declared as an array address
+    fn value_is_array_address(&self, value_id: ValueId) -> bool {
+        self.ssa_value_to_array_address.contains_key(&value_id)
+    }
```

Comment:
Perhaps `dfg.type_of_value(value_id) == Type::Reference` would be more accurate?

Reply:
That wouldn't be able to identify instruction results that encode array address arithmetic.

Reply:
I would think it would, the result of an addition of an address and an offset should itself be an address

Reply:
Right you are. I notice the type is inferred from the binary op's lhs. Is there any danger of the operands being the other way around (and incorrectly returning a numeric constant)?

Reply:
Switched impl